lib/recursive-methods.rb

```diff
@@ -1,49 +1,90 @@
+# Space complexity: O(n) n number of stack frames will be used
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 1 || n == 0
+    return 1
+  elsif n < 0
+    raise ArgumentError
+  end
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the length of s, reverse will get called n/2 times
+# Space complexity: O(n)? where n is the length of s. A new stack frame will be used each call, and by slicing into the array a new array is created during each call, which i think would make it 0(2n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 1 || s.length == 0
+    return s
+  end
+  return s[-1] + reverse(s[1...-1]) + s[0]
+end
+
+# Time complexity: O(n), where n is the length of s, reverse will get called n/2 times
+# Space complexity: O(n) where n is the length of s, n/2 number of stack frames will be used. Unless you are using a language that utilizes tail recursion, then the space complexity would be constant
+def r_reverse_inplace(s, i)
+  temp = s[i]
+  s[i] = s[s.length - 1 - i]
+  s[s.length - 1 - i] = temp
+  return s if i == s.length / 2
+  return r_reverse_inplace(s, i + 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  i = 0
+  return s if s.length == 0
+  return r_reverse_inplace(s, i)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) factorial will be called n number of times, and everything within the function is constant
+# Space complexity: O(n) n number of stack frames will be used
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 1
+    return 2
+  elsif n == 0
+    return 0
+  end
+  return 2 + (bunny(n - 1))
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the length of s, nested will get called n/2 times
+# Space complexity: O(n) where n is the length of s. A new stack frame will be used each call, and by slicing into the array a new array is created during each call, which i think would make it 0(2n)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  matches = { "(" => ")", "{" => "}", "[" => "]" }
+  return true if !s
+  return false if matches[s[0]] != s[-1] || s.length == 1
+  return nested(s[1...-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n),
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+  return false if !array || array.length == 0
+  return true if array[0] == value
+  return search(array[1..-1], value)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the length of s, nested will get called n/2 times
+# Space complexity: O(n) where n is the length of s. A new stack frame will be used each call
 def is_palindrome(s)
```

Comment:
O(n^2) similar to reverse.

Could you have used a technique similar to the one you used with reverse_in_place?